group-coordinator/src/test/java/org/apache/kafka/coordinator/group/generic/GenericGroupTest.java

```diff
@@ -0,0 +1,935 @@
+        joinGroupFuture.whenComplete((joinGroupResult, __) ->
+            joinAwaitingMemberFenced.set(joinGroupResult.errorCode() == Errors.FENCED_INSTANCE_ID.code()));
+        group.add(member, joinGroupFuture);
+
+        AtomicBoolean syncAwaitingMemberFenced = new AtomicBoolean(false);
+        CompletableFuture<SyncGroupResponseData> syncGroupFuture = new CompletableFuture<>();
+        syncGroupFuture.whenComplete((syncGroupResult, __) ->
+            syncAwaitingMemberFenced.set(syncGroupResult.errorCode() == Errors.FENCED_INSTANCE_ID.code()));
+        member.setAwaitingSyncFuture(syncGroupFuture);
+
+        assertTrue(group.isLeader(memberId));
+        assertEquals(memberId, group.staticMemberId(groupInstanceId));
+
+        String newMemberId = "newMemberId";
+        group.replaceStaticMember(groupInstanceId, memberId, newMemberId);
+        assertTrue(group.isLeader(newMemberId));
+        assertEquals(newMemberId, group.staticMemberId(groupInstanceId));
+        assertTrue(joinAwaitingMemberFenced.get());
+        assertTrue(syncAwaitingMemberFenced.get());
+        assertFalse(member.isAwaitingJoin());
+        assertFalse(member.isAwaitingSync());
+    }
+
+    @Test
+    public void testCompleteJoinFuture() {
+        GenericGroupMember member = new GenericGroupMember(
+            memberId,
+            Optional.empty(),
+            clientId,
+            clientHost,
+            rebalanceTimeoutMs,
+            sessionTimeoutMs,
+            protocolType,
+            Collections.singletonList(
+                new Protocol(
+                    "roundrobin",
+                    new byte[0]
+                )
+            )
+        );
+
+        AtomicBoolean invoked = new AtomicBoolean(false);
+        CompletableFuture<JoinGroupResponseData> joinGroupFuture = new CompletableFuture<>();
+        joinGroupFuture.whenComplete((__, ___) ->
+            invoked.set(true));
+        group.add(member, joinGroupFuture);
+
+        assertTrue(group.hasAllMembersJoined());
+        group.completeJoinFuture(member, new JoinGroupResponseData()
+            .setMemberId(member.memberId())
+            .setErrorCode(Errors.NONE.code()));
+
+        assertTrue(invoked.get());
+        assertFalse(member.isAwaitingJoin());
+    }
+
+    @Test
+    public void testNotCompleteJoinFuture() {
+        GenericGroupMember member = new GenericGroupMember(
+            memberId,
+            Optional.empty(),
+            clientId,
+            clientHost,
+            rebalanceTimeoutMs,
+            sessionTimeoutMs,
+            protocolType,
+            Collections.singletonList(
+                new Protocol(
+                    "roundrobin",
+                    new byte[0]
+                )
+            )
+        );
+
+        group.add(member);
+
+        assertFalse(member.isAwaitingJoin());
+        group.completeJoinFuture(member, new JoinGroupResponseData()
```

Comment:
nit: Should we change `completeJoinFuture` to return a boolean? This test is a bit useless otherwise.